### ISSUE TITLE
Incorrect prefer_constructors_over_static_methods warning

### ISSUE THREAD

**Issue by spkersten** [dart-lang/linter#2149]
Consider:

```dart
class Foo {
  Foo._();

  static Foo bar<T>(T value) => Foo._();

  Foo.bar2<T>(T value);
}
```

The static method results in:
`Prefer defining constructors instead of static methods to create instances. (prefer_constructors_over_static_methods`

However, it is not possible to create a constructor with a type parameter.

Dart 2.7.0

**Comment by mraleph** [comment]
Moving to dart-lang/linter - because it is not related to the core SDK.

**Comment by srawlins** [comment]
This issue has more implications:

* A _non-generic_ static method on a generic class _A\<T>_ which returns an A must return an A with a concrete type argument (int, void, dynamic, implicit dynamic, ...), so it cannot be replaced with a constructor.
* A _generic_ static method on a generic class _A\<T>_ which returns an A might use _T_ in its parameter list, so it might not be able to be replaced with a constructor.

I think reporting on either of these cases should be a false positive.

**Comment by bwilkerson** [comment]
> A generic static method on a generic class A<T> which returns an A might use T in its parameter list, so it might not be able to be replaced with a constructor.

This case isn't any different than the preceding non-generic case. Static methods can't refer to the type parameters of the class, so whether the name of the class' type parameter is shadowed makes no difference.

I think the relevant question is whether any instances of `A` are created in the method using a non-type parameter as the type argument to `A`'s constructor. That will always be the cases for non-generic methods because there are no type parameters in scope, but for a generic method this will require analyzing the body of the method.

That said, it isn't clear to me that it's worth the extra effort to compute the answer to that question. Given that it is possible to write a static generic method matching the criteria above but that still couldn't be replaced with a constructor (a false positive), it might be better to just ignore static methods whose return type is the same as the class in which they are declared when the class is generic.

**Comment by srawlins** [comment]
> This case isn't any different than the preceding non-generic case. Static methods can't refer to the type parameters of the class, so whether the name of the class' type parameter is shadowed makes no difference.

Ah good catch.

> That said, it isn't clear to me that it's worth the extra effort to compute the answer to that question.

Yeah I think not.

> Given that it is possible to write a static generic method matching the criteria above but that still couldn't be replaced with a constructor (a false positive), it might be better to just ignore static methods whose return type is the same as the class in which they are declared when the class is generic.

I think that's fine.
